### ISSUE TITLE
Unify bezier animations in the animation player

### ISSUE THREAD

**Issue by fire** [godotengine/godot-proposals#392]
**Describe the project you are working on:**
LillyByte mentioned she had a hard time using the animation player in her project.

https://twitter.com/LillyByteGames

**Describe the problem or limitation you are having in your project:**
Currently the animation player requires you to go to each track to edit the bezier curves.

![image](https://user-images.githubusercontent.com/32321/72675387-dc308100-3a37-11ea-8131-5dcf6437cef7.png)

The animation player only displays one of the nodes.

There is no way to select another node without going back to the timeline list.

**Describe how this feature / enhancement will help you overcome this problem or limitation:**

My proposal is to unifiy all the bezier curve so they're on the same page.

**Show a mock up screenshots/video or a flow diagram explaining how your proposal will work:**

This is my current work in progress.

![ezgif-6-bc34fa6c77da](https://user-images.githubusercontent.com/32321/72675413-785a8800-3a38-11ea-9d63-55e1814a8527.gif)

Akeytsu displays the beziers editor in the viewport:

![image](https://user-images.githubusercontent.com/32321/72675450-13536200-3a39-11ea-9ff6-f745a5d194a4.png)

https://www.youtube.com/watch?v=e8f8xZm9eJY

Maya displays the beziers editor in the properties like Godot Engine: 

![Maya](https://help.autodesk.com/cloudhelp/2018/ENU/Maya-Animation/images/GUID-83E0DFEE-8248-43A7-AE6D-B06098B5458D.gif
)

Looking for suggestions on how to modify my current prototype.

**Describe implementation detail for your proposal (in code), if possible:**

https://github.com/fire/godot/tree/unified-beziers

* Make editor/animation_bezier_editor.h a multiple bezier track data structure [done]
* Make the graph render multiple tracks [done]
* Create a button to open a per animation beziers view [done]
* Modify the animation player so it can work on multiple bezier tracks [done]
* Make beziers view scroll. [pending]
* Modify the bezier view so the developer can select a specific key without switching to another track [pending]
* Modify the animation player so it can work on non bezier tracks [pending / blocked on design]

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
This is a usability issue to improve workflow.

**Is there a reason why this should be core and not an add-on in the asset library?:**
This is a ux modification to the core editor.

**Comment by Master-J** [comment]
This would be a good improvement, i often need to create animations that involve multiple objects and having the possibility to do it without having to switch between windows would be great

**Comment by HungryProton** [comment]
This would be really helpful indeed :+1: 

However displaying all the objects tracks at once all the time might overload the track view.
I really like how Blender does it : Only showing the tracks for the currently selected objects.

With a single one selected :

![image](https://user-images.githubusercontent.com/52043844/72678574-7da2eb80-3aa7-11ea-82ba-ac93f7b0120c.png)

With many objects selected :

![image](https://user-images.githubusercontent.com/52043844/72678357-51866b00-3aa5-11ea-8f08-feebebc60e96.png)

**Comment by LillyByte** [comment]
Yeah, this, and the per-selection idea, would definitely help solve the usability issue with the timeline editor and bezier curves being too far disconnected from one another; making them much more intuitive and user friendly.

**Comment by jitspoe** [comment]
Looking good, fire!  I didn't realize you could even use Bezier curves in animation.  Thought it was just linear or quadratic.  Is this new in 3.2?

**Comment by LillyByte** [comment]
> Looking good, fire! I didn't realize you could even use Bezier curves in animation. Thought it was just linear or quadratic. Is this new in 3.2?

They've been there for quite some time, they've just been really terrible and inconvenient to use to the point that not using them was always the better option.

**Comment by Calinou** [comment]
@jitspoe They were added in 3.1.

**Comment by jitspoe** [comment]
Ah, I see.  You can't change existing animations to Bezier.  That would be nice.  Currently, you just have the nearest/linear/cubic options for existing animations.  Also, adding keyframes for Bezier seems to be bugged.  Instead of adding a new keyframe, it adds a whole new track:

![image](https://user-images.githubusercontent.com/7751182/73601673-a610f800-4533-11ea-8a61-257c9edf5a69.png)

Adding non-Bezier keyframes adds another frame to the existing track.  Maybe that's why I never used it, since it just doesn't work. :)

**Comment by sketchyfun** [comment]
The ability to convert existing keyframes to bezier would be great, or even just have bezier as the default for anything other than 'Nearest'. 

After all, linear is the same as bezier without the in/out handles, and cubic is basically bezier with less control. Pretty much all the programs I can think of that have animation graphs use a single type of timeline, and the keyframes can be customised for different interpolation. Something like the way Blender does it maybe?

![image](https://user-images.githubusercontent.com/10464013/112486024-6c243600-8d73-11eb-9157-331eaa539392.png)

**Comment by TokageItLab** [comment]
In After Effects, each track has a checkbox for displaying it in the graph editor, and the zoom value immediately after switching to the graph editor is adjusted to the maximum possible value (ex. float use HINT_RANGE) among the checked ones. Also, for values whose range is defined by HINT_RANGE, the overshooting value should be cropped.

**Comment by Dancovich** [comment]
> After all, linear is the same as bezier without the in/out handles, and cubic is basically bezier with less control. Pretty much all the programs I can think of that have animation graphs use a single type of timeline, and the keyframes can be customised for different interpolation. Something like the way Blender does it maybe?

Just watched a previous Unity developer use Godot for his LD54 entry and that caused a lot of trouble for him. He is pixel artist and animator first and developer second, so he's quite used to animation tools and Godot animator editor did not impress him.

Your suggestion of making it like Blender makes sense for me. Instead of a global interpolation mode for the entire track, all tracks are just bezier curves but at each section you can choose the key and mark it as linear, cubic or "custom", where you just edit the bezier curve.

**Comment by fire** [comment]
I am not able to work on this, but anyone who is interested in unifying bezier animations is able to lead on it and I can assist.

**Comment by fire** [comment]
Some work in Add a way to convert animation to Bezier Curve trackers #9597

**Comment by Saul2022** [comment]
Being curious what's the status of this ? I ask since it would be a great addition

**Comment by TokageItLab** [comment]
Recently, the UI itself has been improving, so I'll comment here on a proposal regarding the data structure.

In my opinion, Bezier tracks should be deprecated in the future and migrated to become one of the InterpolationType options for Value tracks. One reason is that we need to resolve the issue of Bezier tracks utilizing subpaths of the Variant type. In other words, the goal is to resolve https://github.com/godotengine/godot/issues/49431 by more semantic way:

Currently blendable:
```
ValueTrack: Node2D/position:x
BezierTrack: Node2D/position:x
```

Currently not blendable:
```
ValueTrack: Node2D/position
BezierTrack: Node2D/position:x
```

So I suggest it migrate into below and deprecate numerical elements of Variant subpaths:
```
ValueTrack: Node2D/position
BezierTrack: Node2D/position
```

Furthermore, since currently calculating Bezier curves at runtime is extremely inefficient, you should bake the curves immediately after editing and use the baked segments at runtime.

However, I anticipate that a converter for migration will be essential at that time. This means we might need to wait for something like Godot 5.
